File: misc/compile-time/pad.pl
use Data::Dumper;
use strict;

# this is the algorithm for keeping the compile-time environment in pure-perl
# incrementally set environment; and keep a pad stack


my @prog = (

    q!  my $x = 3       !,

    q!  print "x=$x\n"  !,

    q!  my $y = 4;      !,

    q!  print "y=$y\n"  !,

    q!  my $z = 7;      !,

    q!  $y++            !,

    q!  print "y=$y\n"  !,

    q!  my $k = $y + 1; print "k=", $k, "\n"  !,

);

my @vars;
sub get_vars {
    my $s = shift;
    push @vars, $s =~ /([\$\@\%]\w+)/;
}

my $pad;
while (@prog) {

    my $line = shift(@prog);
    get_vars($line);
    my $cmd =
          'do { '
        .   $line . '; '
        .   'sub { '
        .       join(',', @vars) . '; '
        .       'eval $_[0] '
        .   '} '
        . '}';

    if ($pad) {
        $pad = $pad->($cmd)
    }
    else {
        $pad = eval $cmd;
    }

    #print "cmd $cmd\n";
    #print "pad $pad - $@\n";

}

File: misc/compile-time/pad.js

// this is the algorithm for keeping the compile-time environment 
// incrementally set environment; and keep a pad stack

function print (s) { process.stdout.write(s) }

var prog = [

    '  var x = 3       ',

    '  print( "x=" + x + "\\n")  ',

    '  var y = 4;      ',

    '  print( "y=" + y + "\\n")  ',

    '  var z = 7;      ',

    '  y++            ',

    '  print( "y=" + y + "\\n")  ',

    '  var k = y + 1; print( "k="+ k+ "\\n")  '

];

var pad;
while (prog.length) {
    var line = prog.shift();
    cmd = '(function () { ' + 
                line + '; ' + 
                'return function (cmd) { ' + 
                    'return( eval(cmd) ) ' + 
                '} ' + 
           '}())';
    if (pad) {
        pad = pad(cmd)
    }
    else {
        pad = eval(cmd);
    }
}

